lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -1416,53 +1416,111 @@ static EnableOp getLastEnableOp(SeqOp parent) {
-///
-///   if %a with %A {                       if %a with %A {
-///     seq { ... calyx.enable @B }           seq { ... }
-///   else {                          ->    } else {
-///     seq { ... calyx.enable @B }           seq { ... }
-///   }                                     }
-///                                         calyx.enable @B
-static LogicalResult eliminateCommonTailEnable(IfOp ifOp,
-                                               PatternRewriter &rewriter) {
-  // Check if the branches exist.
+/// Returns a mapping of {Enabled Group, EnableOp} for all EnableOps within the
+/// immediate ParOp's body.
+static llvm::StringMap<EnableOp> getAllEnableOpsInImmediateBody(ParOp parent) {
+  llvm::StringMap<EnableOp> enables;
+  Block *body = parent.getBody();
+  for (EnableOp op : body->getOps<EnableOp>())
+    enables.insert(std::pair(op.groupName(), op));
+
+  return enables;
+}
+
+/// Removes common tail enable operations for then and else regions encompassed
+/// in SeqOp and ParOp. This canonicalization is stringent about not entering
+/// nested control operations, as this may cause unintentional changes in
+/// behavior.
+static LogicalResult eliminateCommonTail(IfOp ifOp, PatternRewriter &rewriter) {
   if (!ifOp.thenRegionExists() || !ifOp.elseRegionExists())
     return failure();
+  rewriter.setInsertionPointAfter(ifOp);
 
-  auto &thenOpStructureOp = ifOp.getThenBody()->front();
-  auto &elseOpStructureOp = ifOp.getElseBody()->front();
-  // TODO(circt/#1861): ParOps have less restrictive conditions.
-  if (isa<ParOp>(thenOpStructureOp) || isa<ParOp>(elseOpStructureOp))
-    return failure();
-
-  // At this point, only sequential operations are valid inside the branches.
-  auto thenSeqOp = dyn_cast<SeqOp>(thenOpStructureOp);
-  auto elseSeqOp = dyn_cast<SeqOp>(elseOpStructureOp);
-  assert(thenSeqOp && elseSeqOp &&
-         "expected nested seq ops in both branches of a calyx.IfOp");
-
-  EnableOp lastThenEnableOp = getLastEnableOp(thenSeqOp);
-  EnableOp lastElseEnableOp = getLastEnableOp(elseSeqOp);
+  auto &thenControl = ifOp.getThenBody()->front();
+  auto &elseControl = ifOp.getElseBody()->front();
+  if (auto r1 = dyn_cast<ParOp>(thenControl);
+      auto r2 = dyn_cast<ParOp>(elseControl)) {
+    ///   if %a with @G {                  if %a with @G {
+    ///     par {                            par { ... }
+    ///       ...                          } else {
+    ///       calyx.enable @A     ->         par { ... }
+    ///       calyx.enable @B              }
+    ///     }                              calyx.enable @A
+    ///   else {                           calyx.enable @B
+    ///     par {
+    ///       ...
+    ///       calyx.enable @A
+    ///       calyx.enable @B
+    ///     }
+    ///   }
+    llvm::StringMap<EnableOp> A = getAllEnableOpsInImmediateBody(r1),
+                              B = getAllEnableOpsInImmediateBody(r2);
+
+    // Compute the intersection between `A` and `B`.
```

Comment:
I assume that the intersection has been manually calculated due to the map-like structure of A and B, preventing the use of `std::set_intersection`. Is there not a way to use the keys of each map (vector-like) to then do a set intersection with?

Reply:
Hmm I should try `set_intersection` or LLVM's version again. Before I had something more complicated written.

Reply:
So on second thought, this isn't possible since we essentially want set intersection with side effects. 

We could use the keys of each map to perform a set_intersection, then iterate through `A` and `B` and delete keys with the given `groupName`, but I'm not convinced that's any cleaner than this. WDYT?

Reply:
If it'll complicating the code just for the sake of shoehorning an stl function in there, then that probably doesn't make that much sense, so up to you.

Reply:
Given this is only used once, I don't think its necessary. Perhaps in the future we can right some wrapper for set intersection with side effects. Is there anything else blocking your approval?